In [1]:
from sklearn.datasets import load_iris
from sklearn.metrics import accuracy_score
from PerceptronLearningRule.Perceptron import *
import random

In [2]:
random.seed(1754339)

# 3c Iris dataset

In [3]:
iris = load_iris()
targetNames = iris['target_names']
target = iris['target']
data = iris['data']

### Classificeer eerst alleen de types Setosa en Versicolour.

In [4]:
SetosaAndVersicolourData = data[:100]
SetosaAndVersicolourTarget = target[:100]

In [5]:
perceptron = Perceptron([random.randrange(-20, 20), random.randrange(-20, 20), random.randrange(-20, 20), random.randrange(-20, 20)], random.randrange(-20, 20))
print("Staring perceptron:", perceptron)

Staring perceptron: input: [], weights: [9, -7, 8, -11], bias: 3, output: None


#### Het trainen van de perceptron op de Setosa en Versicolour data

In [6]:
for x in range(10):
    for i in range(len(SetosaAndVersicolourData)):
        perceptron.setInput(list(SetosaAndVersicolourData[i]))
        perceptron.update(SetosaAndVersicolourTarget[i], learningRate=0.8)

In [7]:
print("fitted perceptron: ", perceptron)

fitted perceptron:  input: [5.7, 2.8, 4.1, 1.3], weights: [3.5600000000000005, -12.680000000000001, 10.88, -9.32], bias: 1.4000000000000001, output: 1


Uiteindelijke parameters:

Weights: [3.5600000000000005, -12.680000000000001, 10.88, -9.32]

Bias: 1.4000000000000001

#### Het predicten van de 2 bloemen met de perceptron

In [8]:
pred = []
for i in range(len(SetosaAndVersicolourData)):
    perceptron.setInput(list(SetosaAndVersicolourData[i]))
    perceptron.run()
    pred.append(perceptron.output)

In [9]:
accuracy_score(SetosaAndVersicolourTarget, pred) * 100

100.0

Een mooie score van 100%, wel overfitted ;)
Maar het gaat natuurlijk om de functionaliteit van het leren

### Classificeer daarna alle drie de types, dus ook Verginica

In [10]:
perceptron = Perceptron([random.randrange(-20, 20), random.randrange(-20, 20), random.randrange(-20, 20), random.randrange(-20, 20)], random.randrange(-20, 20))
print("staring perceptron:", perceptron)

staring perceptron: input: [], weights: [18, -4, -3, 1], bias: 8, output: None


#### Het trainen van de perceptron op alle data

In [11]:
for x in range(10):
    for i in range(len(data)):
        perceptron.setInput(list(data[i]))
        perceptron.update(target[i], learningRate=0.8)

In [12]:
print("fitted perceptron: \n", perceptron)

fitted perceptron: 
 input: [5.9, 3.0, 5.1, 1.8], weights: [844.8800000000007, -53.6800000000001, 1691.9600000000046, 723.4799999999992], bias: 46.39999999999997, output: 1


Uiteindelijke parameters:

Weights: [837.5600000000003, -47.71999999999999, 1687.1200000000042, 740.8000000000004]

Bias: 46.999999999999964

#### Het predicten van de 3 bloemen met de perceptron

In [13]:
pred = []
for i in range(len(data)):
    perceptron.setInput(list(data[i]))
    perceptron.run()
    pred.append(perceptron.output)

In [14]:
accuracy_score(target, pred) * 100

33.33333333333333

Waarom krijg je met 3 bloemmen zo'n slechte score?
Dit komt omdat een enkele perceptron alleen met lineare problemen kan omgaan.
3 mogelijke uitkomsten zijn helaas niet linear. Hiervoor is een netwerk van Sigmoid neuron nodig.
Omdat het niet mogelijk is om met de step functie een netwerk te leren door backpropagation, hiervoor is echt de sigmoid functie vereist.